date_range/models/date_range_type.py

```diff
@@ -42,3 +71,42 @@ def _check_company_id(self):
                         _('You cannot change the company, as this '
                           'Date Range Type is  assigned to Date Range '
                           '(%s).') % (rec.date_range_ids.name_get()[0][1]))
+
+    @api.depends("date_range_ids")
+    def _compute_date_ranges_exist(self):
+        for dr_type in self:
+            dr_type.date_ranges_exist = bool(dr_type.date_range_ids)
+
+    @api.onchange("name_expr")
+    def onchange_name_expr(self):
+        """Wipe the prefix if an expression is entered.
+
+        The reverse is not implemented because we don't want to wipe the
+        users' painstakingly crafted expressions by accident.
+        """
+        if self.name_expr and self.name_prefix:
+            self.name_prefix = False
+
+    def test_name_expr(self):
+        """Button on the form to preview newly generated names"""
+        self.ensure_one()
+        year_start = fields.Datetime.now().replace(day=1, month=1)
+        next_year = year_start + relativedelta(years=1)
+        names = self.env["date.range.generator"]._generate_names(
+            [year_start, next_year], self.name_expr, self.name_prefix)
+        raise UserError("\n".join(names))
+
+    @api.model
+    def autogenerate_ranges(self):
+        """Generate ranges for types with autogeneration settings"""
+        for dr_type in self.search(
+                [("autogeneration_count", "!=", False),
+                 ("autogeneration_unit", "!=", False)]):
+            try:
+                with self.env.cr.savepoint():
+                    form = Form(self.env["date.range.generator"])
+                    form.type_id = dr_type
+                    wizard = form.save()
+                    wizard.action_apply()
+            except (AssertionError, UserError, ValidationError):
+                pass
```

Comment:
We should at least log an exception...

Reply:
I considered that, but those errors are usually related to no new periods needed to be created.

Reply:
Can we just ignore the specific error "No ranges to generate with these settings" ? You could create a typed error for this one only that inherits from UserError to ease to ignore it. Or you could pass a paramter to the action_apply method to ignore this specific error... What do you think about?

Reply:
Improved the error handling and ignored 'No ranges to generate' message in batch mode. Thank you for the suggestion!